### DISCUSSION TITLE
SSR support for Next.js 13 `/app` directory?

### DISCUSSION THREAD

**Discussion by pruthvi145** [algolia/instantsearch#5413] [General] [answered]
Hi Team, do you have any plan to support [Next.js 13](https://beta.nextjs.org/docs), specially for the SSR?

I tested for the client-side rendering, seems working so far. Few routing issues but those are already addressed [here](https://github.com/algolia/instantsearch/issues/5241).

However, SSR is not working at all with the [`/app` directory (beta)](https://beta.nextjs.org/docs/app-directory-roadmap). Feels like need the changes in the `react-instantsearch-hooks-server` package.

**Comment by Haroenv** [comment]
We haven't yet tried the beta app directory. At the moment I'm not sure where it would fit in, as they expect usage of `fetch` directly, while we use our own API client, and it requires knowing which widgets will be rendered in a two-pass process.

**Comment by philwolstenholme** [comment]
We (http://shop.coop.co.uk) are starting to rebuild our site in Next 13 and we would love to see support for the new `app` router. At the moment it looks like we'll be using the `app` directory for the majority of our site, and it will be a bit inconvenient to have to also set up page layouts etc to work with the `pages` directory just for InstantSearch.

@Haroenv it might be worth engaging with some of the [Next/Vercel developer relations team](https://vercel-team.vercel.app/) (led by @leerob) to see if they can give any advice on your query/thought above.

I also noticed that the Next 13 `app` router docs have a section on data fetching without fetch (it's not mandatory): https://beta.nextjs.org/docs/data-fetching/fetching#data-fetching-without-fetch and there is also the `use` hook (https://beta.nextjs.org/docs/data-fetching/fetching#use-in-client-components) that can be used on the client-side to play nicely with React Server Components.

**Comment by Colton** [comment]
My team would also benefit from support for SSR in the `/app` directory. I would be happy to contribute as well.

**Comment by christiankaindl** [comment]
The [`react-instantsearch-hooks-router-nextjs`](https://github.com/algolia/instantsearch/tree/master/packages/react-instantsearch-hooks-router-nextjs) package also doesn't support the NextJS 13 app directory at the moment. Is there a roadmap for app dir compatibility?

**Reply by aymeric-giraudet** [reply]
From what I've experimented, it seems the app router does not rely on having `__N` is history state anymore, so I don't think it'll be needed. Back button should work out of the box.

**Reply by jmotes** [reply]
@aymeric-giraudet do you know if is this still true? I'm working on migrating from pages to the app router but we're using a purely clientside approach still so I'm hoping this is the case!

**Reply by jmotes** [reply]
~Update: I tried it and it doesn't work - the back button still clears out the previous filter with Next.js 14.2.4 + app router.~

~I'll try to get the experimental solution working...~

Our custom router doesn't work, but simplifying it with the newer stateMapping methods seems to be working fine!

**Comment by philwolstenholme** [comment]
The `app` router (previously known as the `app` directory) in Next.js just got made stable, so it's now a stable target to start working on, and I imagine adoption in Algolia's customers will increase: https://nextjs.org/blog/next-13-4

**Comment by Colton** [comment]
Hey all,

Just wanted to bump this issue again. We worked around this by going client-side but I'd like to know if there's any plans for addressing this on the roadmap. Thanks!

**Comment by naseef0** [comment]
Any updates on this?

I'm also waiting for a proper solution with Next js 13 app dir.

**Comment by tilman** [comment]
@Haroenv can you give an update on this topic? Is it on your roadmap or can we expect to build a solution on our own with direct API access if we want to realize a app-router project within the next half year?

**Reply by Haroenv** [reply]
https://github.com/algolia/instantsearch/discussions/5413#discussioncomment-6427443

**Comment by Haroenv** [comment]
While `use` is not yet stable, we do have an example on how to use React InstantSearch Hooks with React Server Components.

There are some caveats that prevent us from documenting this officially
- React's use is not stable, sometimes causing infinite renders
- access to the router isn't possible / straightforward in Next App dir
- all code (including renderToString) is bundled for the client as well
- in some cases getServerState is executed client-side, which isn't needed and can cause more network requests than needed


Those caveats taken in account, here's an example: https://codesandbox.io/p/sandbox/interesting-rhodes-4pgtu6?file=%2Fapp%2Fpage.tsx

```tsx
"use client";

import React, { use } from "react";
import algoliasearch from "algoliasearch/lite";
import { Hit as AlgoliaHit } from "instantsearch.js";
import {
  DynamicWidgets,
  InstantSearch,
  Hits,
  Highlight,
  RefinementList,
  SearchBox,
  InstantSearchSSRProvider,
  InstantSearchServerState,
} from "react-instantsearch-hooks-web";
import { renderToString } from "react-dom/server";
import { Panel } from "../components/Panel";
import { getServerState } from "react-instantsearch-hooks-server";

const client = algoliasearch("latency", "6be0576ff61c053d5f9a3225e2a90f76");

type HitProps = {
  hit: AlgoliaHit<{
    name: string;
    price: number;
  }>;
};

function Hit({ hit }: HitProps) {
  return (
    <>
      <Highlight hit={hit} attribute="name" className="Hit-label" />
      <span className="Hit-price">${hit.price}</span>
    </>
  );
}

type SearchPageProps = {
  serverState?: InstantSearchServerState;
};

function Search({ serverState }: SearchPageProps) {
  return (
    <InstantSearchSSRProvider {...serverState}>
      <InstantSearch
        searchClient={client}
        indexName="instant_search"
        insights={true}
      >
        <div className="Container">
          <div>
            <DynamicWidgets fallbackComponent={FallbackComponent} />
          </div>
          <div>
            <SearchBox />
            <Hits hitComponent={Hit} />
          </div>
        </div>
      </InstantSearch>
    </InstantSearchSSRProvider>
  );
}

function FallbackComponent({ attribute }: { attribute: string }) {
  return (
    <Panel header={attribute}>
      <RefinementList attribute={attribute} />
    </Panel>
  );
}

export default function SearchPage() {
  const serverState = use(
    getServerState(<Search />, {
      renderToString,
    })
  );

  return <Search serverState={serverState} />;
}
```

Remember as well that when you use App dir, you don't have to convert your entire app at once, you can leave the search page as a regular "pages" component.

**Reply by destino92** [reply]
Can you add routing to this snippet please?

**Reply by aymeric-giraudet** [reply]
Hi @destino92 you can see it in this comment : https://github.com/algolia/instantsearch/discussions/5413#discussioncomment-6508691

Although it's still pretty unstable because of `use`

**Comment by philwolstenholme** [comment]
> Remember as well that when you use App dir, you don't have to convert your entire app at once, you can leave the search page as a regular "pages" component.

This is what we're doing at the moment for our search page and category/taxonomy page, but it's not ideal – there is a hard browser navigation/load as users move between the `app` router and `pages` router, and it also means some duplication of our layout approach to support both router types, remembering to wrap both `app` and `pages` with the right context providers when we add new providers, etc.

Thanks a lot for sharing the experimental example, it's great to see that your team have been looking into this :)

**Comment by timmeade** [comment]
Not really an option for us as we are converting to new app dir in next js and going backwards to the pages folder just is not an option.  We have instantsearch working under new structure except for routing section.   And nothing we try seems to work.  

Could be the algolia deal breaker.    Seems we would have a working option by now.  It's been months.  Answer above is not "answer" as it does not contain routing.

**Reply by aymeric-giraudet** [reply]
Hey @timmeade,

I've had a look and got something that kinda works : https://codesandbox.io/p/sandbox/react-instantsearch-hooks-next-app-dir-forked-qffqwj?file=/app/Search.tsx:11,18

For some reason the URL doesn't get updated in the CodeSandbox preview but you can see it in action directly on the URL : https://qffqwj-3000.csb.app/?instant_search%5BrefinementList%5D%5Bbrand%5D%5B0%5D=Apple

@Haroenv's answer still stands though, as of newer Next.js versions I seldom get an error saying `async/await` is not supported in client components yet, which is [indeed mentioned in their documentation](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching#use-in-client-components).

As long as it is the case, we cannot provide a simple solution that doesn't involve radically changing how the lib operates.

We're currently trying to reach out to the DevRel team of Next.js to find a solution, whether it is having to wait for proper support of `async/await` in client components, or a way to parse which InstantSearch widgets are to be mounted from a server component.

A lot of third-party libs are having trouble with the new `app` dir, and it's even more difficult with a lib that does UI, state management, routing and network calls altogether.

I hope this example can help you while we are actively trying to properly support the new `app` dir.

**Reply by philwolstenholme** [reply]
@leerob - sending a big +1 from this Vercel customer and Next user for seeing if the Vercel devrel team can help out the Algolia team with their `app` router implementation, please and thank you!🙏

**Reply by timmeade** [reply]
@leerob another +1.   @aymeric-giraudet  Thanks for the work.   The CSB looks good.   Let me dig through it and see if we can make it work.

**Reply by timmeade** [reply]
@aymeric-giraudet maybe you can help.   

We have it working except we are using a CustomHits component.   On the old pages version when an item was clicked using a Link tag the window closed.

Under the app dir when the item is clicked the url does change and new pages load, but the window/modal opened in CustomHits stays in place on top.   The data under the open search actually changes.  I think if we were to 'clear' the current search during the click on Link that it would go away.  Any of this ring a bell?

Thanks again

**Reply by timmeade** [reply]
The sandbox above does not work in code.


```
  const serverState = use(
    getServerState(<Search serverUrl={serverUrl} />, {
      renderToString,
    })
  );
```

triggers the 

Unhandled Runtime Error
Error: async/await is not yet supported in Client Components, only Server Components. This error is often caused by accidentally adding `'use client'` to a module that was originally written for the server.

**Reply by destino92** [reply]
@leerob +1 please

**Comment by aymeric-giraudet** [comment, marked as answer]
Hey all, I'm updating this discussion with news !

We released an **experimental** package which supports Next.js App Router, you can check its usage there : https://www.algolia.com/doc/guides/building-search-ui/going-further/server-side-rendering/react/#app-router-experimental

Don't hesitate to provide feedback right here !

A few drawbacks we already know of are that using the browser's forward button from one InstantSearch page to another can have issues, and there currently is a useless query being made on the client but it will be fixed by https://github.com/algolia/instantsearch/pull/5854

**Reply by Colton** [reply]
Woohoo! Thanks for you and your team's hard work @aymeric-giraudet !

I can't seem to get hooks working in the App Router with this new guide. I have tried putting the <InstantSearchNext> HOC in both my component and in the layout as a whole and I still end up getting the "[InstantSearch] Hooks must be used inside the <InstantSearch> component." error.

Here's a snippet as an example...

```javascript
export function Search() {
  const { hits } = useHits()
  return (
    <>
      <SearchBox />
    </>
  )
}
```

**EDIT:** I fixed the issue. I was importing `useHits()` from `react-instantsearch-hooks-web`. When I switched the import to `react-instantsearch`, it worked. Apologies for the oversight. Thanks again for the great work!

**Reply by omerm1** [reply]
i still get this issue, when trying to use useInstantSearch(), could i see more of your code to help me out please?

**Reply by Colton** [reply]
@omerm1 is the component in which you have `useInstantSearch()` wrapped in the `<InstantSearchNext>` HOC?

My setup is very simple. The `Search` component in the original comment is simply wrapped in the HOC via my root layout.

```javascript
 const InstantSearchProvider = ({ children }: { children: React.ReactNode }) => (
      <InstantSearchNext indexName={index} searchClient={searchClient} routing>
        {children}
      </InstantSearchNext>
)
```

Seeing _your_ code would be more of a help.

**Reply by wesleyjanse** [reply]
Using category pages (initial filter using Configure), on routing the server render sometimes renders 0 items first before the client taking over and showing the content itself. 

Same with using refinements/dynamicRefinements, the filtering doesn't seem to apply on initial render. 

Any clue why this is happening?

I have a very simple setup with just the HOC, which allows children, i use the HOC, and pass in the correct children, ie useHits, useDynamicWidgets and render the facets using those attributes. (And some other hooks, pagination, sorts etc...)

**Reply by Haroenv** [reply]
@wesleyjanse, I'd recommend you create a fresh issue with reproduction. I don't see on first idea what the cause could be (maybe routing doesn't get taken in account, maybe widgets are not mounted, maybe ui state somehow gets lost, maybe the HOC causes its children not to be mounted on server)

**Reply by buesing** [reply]
I'm using `react-instantsearch@7.3.0` with `react-instantsearch-nextjs@0.1.4` and `next@13.5.6` and the experience is pretty buggy. I'm using various hooks scattered across the app (`useClearRefinements`,  `useInstantSearch`, `useRefinementList`, `useRange`) and the `useRefinementList` are mostly not getting any data on the server (returning an empty array). I'm noticing that as I'm removing other hooks calls, it starts to work. For example if I reduce the app to just one call to `useRefinementList`, I'm seeing data on the server. But as I add more hook calls to the component body, it starts to return an empty array.
Any hints as to what this could be? I don't have the resources right now to make a full reproduction repo, just thought I'd share my experience.

**Reply by Haroenv** [reply]
I don't have any awareness of any of those issues @buesing, if you can reproduce and open a new issue, we can investigate what's happening in your case.

**Reply by aymeric-giraudet** [reply]
@buesing 

This could be due to `useInstantSearch`, do you try to access the results from it ?
Also do you conditionally render certain hooks ?

It's virtually impossible for us to find an answer without a reproduction, it doesn't need to be a full-fledged application, just a small CodeSandbox reproduction would do !

**Reply by buesing** [reply]
Hello again @Haroenv @aymeric-giraudet, I have managed to make a super simple reproduction repo: https://github.com/buesing/instantsearch-next-repro
If you clone and start it, you will see that the log of the `surroundings` is empty on the server and in the initial render. It only seems to load in the client. However, removing one of the 2 calls to `useRefinementList` seems to fix it - then the data is already existent on the server render.

**Reply by aymeric-giraudet** [reply]
Hey @buesing, thanks a ton for the reproduction !

This is because we're using the experimental [`use`](https://react.dev/reference/react/use) hook which allows pausing and resuming the render of **components**.

Here you have a component that calls 2 `useRefinementList` in succession, the problem is that the first one being paused, the second won't be called so the request won't be made with that second facet :/

`use` is still experimental and that's a limitation I didn't realize would be a problem. The only way this can be avoided for now would be to only call our hooks once per component. 

I'll see with the React team if it's possible to have `use` only suspend a hook rather than the entire component.

**Reply by buesing** [reply]
OK, that's a big limitation, but good to know anyway, then I can work around that. Thanks!

**Reply by aymeric-giraudet** [reply]
@buesing 

If React 19 releases with `use` working as it does, i.e. pausing the component rather than hook, I think we could provide an option to have our hooks such as `useRefinementList` not call `use` themselves, and provide a hook like `useWaitForResults` that you could call after all these hooks. Can't think of another alternative for now but at least this would work.

Thanks again for your feedback, we'll make sure to address this once we release the stable version !

**Reply by LuudJanssen** [reply]
Hi @aymeric-giraudet, another alternative might be to work towards a more promise-based API, where the developer is in control of using `use` in the component itself (and also prevent possible waterfalls). This would create a more bloated syntax, but as a tradeoff for flexibility. Example:

```tsx
"use client";

function SearchResults() {
  const { query, resultsRequest } = useInstantSearch() // resultsRequest is just a promise
  const otherWorkRequest = doSomeOtherWork(query) // also returns a promise
  
  const [results, workResponse] = use(Promise.all([resultsRequest, otherWorkRequest]))
  
  // ...
}
```

I'm not convinced this is the right approach, but I would like the flexibility of preventing a request waterfall (if `useInstantSearch()` would suspend, `doSomeOtherWork` wouldn't be called until after the suspension of `useInstantSearch()` was resolved).

**Reply by Haroenv** [reply]
@LuudJanssen, that may make sense for the initial query on server, but likely wouldn't work for any subsequent requests, where the promise shouldn't prevent rendering (otherwise you can't do optimistic UI)

**Reply by LuudJanssen** [reply]
Hi @Haroenv! If you want to do optimistic UI, `resultsQuery` could be a resolved promise. I believe `use()` won't suspend if the promise is already resolved. Am I understanding you correctly?

**Reply by nomadscode** [reply]
@aymeric-giraudet I have the same issue as @wesleyjanse is experiencing. Also, when I load my page that contains the InstantSearchNext directly, I don't get any warnings. But if I start at a different page (like my homepage) and browse to or switch between the page with InstantSearchNext, I get a mounted/unmounted warning. I gets pretty messy with the multiple renders of results.

Guess that's the same issue others are experiencing. Is there a way around it for now?

**Reply by tflx** [reply]
Can we somehow remove the console.log in dev mode?…It's pretty annoying.
